File: backend-spring-boot/src/main/java/no/vicx/backend/user/validation/ProfileImage.java
package no.vicx.backend.user.validation;

import jakarta.validation.Constraint;
import jakarta.validation.Payload;

import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

@Constraint(validatedBy = ProfileImageValidator.class)
@Target(ElementType.PARAMETER)
@Retention(RetentionPolicy.RUNTIME)
public @interface ProfileImage {
    String message() default "";
    String invalidFileTypeMessage() default "{vicx.constraints.ProfileImage.type.message}";
    String invalidSizeMessage() default "{vicx.constraints.ProfileImage.size.message}";
    long maxFileSize() default 50 * 1024; // Default: 50KB

    Class<?>[] groups() default {};

    Class<? extends Payload>[] payload() default {};
}

File: backend-spring-boot/src/main/kotlin/no/vicx/backend/user/validation/ProfileImage.kt
package no.vicx.backend.user.validation

import jakarta.validation.Constraint
import jakarta.validation.Payload
import kotlin.reflect.KClass


@Constraint(validatedBy = [ProfileImageValidator::class])
@Target(AnnotationTarget.VALUE_PARAMETER)
@Retention(AnnotationRetention.RUNTIME)
annotation class ProfileImage(
    val message: String = "",
    val invalidFileTypeMessage: String = "{vicx.constraints.ProfileImage.type.message}",
    val invalidSizeMessage: String = "{vicx.constraints.ProfileImage.size.message}",
    val maxFileSize: Long = (50 * 1024).toLong(), // Default: 50KB
    val groups: Array<KClass<*>> = [],
    val payload: Array<KClass<out Payload>> = []
)
